documentation/projects/proposals/trust_and_safety/detecting_sensitive_textual_content/20230506-implementation_plan_frontend_blurring_sensitive.md

```diff
@@ -0,0 +1,264 @@
+
+## Scope
+
+<!-- An overview of the implementation plan, if necessary. Save any specific steps for the section(s) below. -->
+
+This plan describes the approach, particularly pertaining to the frontend, for
+blurring sensitive content.
+
+The content safety project has an
+[API side](/projects/proposals/trust_and_safety/detecting_sensitive_textual_content/20230309-project_proposal_detecting_sensitive_textual_content.md#api),
+extensively documented in the project proposal but out of the scope for this IP.
+For now, we document what the frontend needs and assume that these needs will be
+met by the API.
+
+## Expected Outcomes
+
+<!-- List any succinct expected products from this implementation plan. -->
+
+The frontend should not show the user any sensitive material unless the user has
+explicitly consented to be exposed to it. Exposure to sensitive material will be
+opt-in, meaning that it will be concealed by default.
+
+## Outlined Steps
+
+<!-- Describe the implementation step necessary for completion. -->
+
+### Step 1: Frontend components/features
+
+The implementation would require the following frontend components to be
+developed.
+
+- Blurrable search result components
+
+  - image cell
+  - audio (box and row layouts)
+
+- Single result page
+
+  - Content safety wall
+  - Hide content button
+
+#### Blur mechanism
+
+For the purposes of faster development, we will go with a CSS blur based
+approach for this implementation plan. It has the distinct advantage over all
+other approaches of being super generalised meaning that it will satisfy the
+blurring needs of images (blurring the image), audio (blurring the title) and
+even future media types like 3D models (blurring the JS-based preview).
+
+### Step 2: Feature flag system enhancements
+
+The feature flag + preferences mechanism will need some new features to support
+some of the new use cases proposed for this feature.
+
+- Ability to store and retrieve preferences from `sessionStorage` instead of
+  `localStorage`.
+- Option to not support `ff_` query parameters for a particular flag.
+
+These features are needed to support the new feature flag
+[`fetch_sensitive`](#fetch_sensitive).
+
+### Step 3: New feature flags
+
+```{caution}
+Observe the difference between on/off and enabled/disabled. The former refers to
+the position/value of the toggle, while the latter refers to whether the toggle
+can be interacted with and flipped.
+```
+
+Two feature flags need to be developed for this feature. One related feature
+flag, `fake_sensitive`, already exists.
+
+#### `fetch_sensitive`
+
+This is not technically a feature flag, but rather a preference, which is
+internally the same API. As a preference, it will always be set to 'switchable'
+in all environments, defaulting to 'off', and will be controlled by the user via
+the switches in the search results sidebar.
+
+This flag will not support the use of `ff_` query parameters, to prevent
+URL-based malicious action.
```

Comment:
This is exciting! It's cool that the same 'feature flag' approach you developed can also be used for these kinds of use-cases as well :tada:

Reply:
Yeah, it was not a use-case when we developed the flags but it turned out that the analytics feature flag, if left switchable, could work as a user-preference. Eventually we added as an official use-case to the [feature-flags documentation](https://docs.openverse.org/frontend/reference/feature_flags.html).